go/vt/vtgate/engine/fk_cascade.go

```diff
@@ -0,0 +1,180 @@
+/*
+Copyright 2023 The Vitess Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package engine
+
+import (
+	"context"
+
+	"vitess.io/vitess/go/sqltypes"
+	querypb "vitess.io/vitess/go/vt/proto/query"
+	vtrpcpb "vitess.io/vitess/go/vt/proto/vtrpc"
+	"vitess.io/vitess/go/vt/vterrors"
+)
+
+type FkChild struct {
+	BVName string
+	Cols   []int // indexes
+	Exec   Primitive
+}
+
+// FkCascade is a primitive that implements foreign key cascading using Selection as values required to execute the FkChild Primitives.
+// On success, it executes the Parent Primitive.
+type FkCascade struct {
+	Selection Primitive
+	Children  []FkChild
+	Parent    Primitive
+
+	txNeeded
+}
+
+// RouteType implements the Primitive interface.
+func (fkc *FkCascade) RouteType() string {
+	return "FkCascade"
+}
+
+// GetKeyspaceName implements the Primitive interface.
+func (fkc *FkCascade) GetKeyspaceName() string {
+	return fkc.Parent.GetKeyspaceName()
+}
+
+// GetTableName implements the Primitive interface.
+func (fkc *FkCascade) GetTableName() string {
+	return fkc.Parent.GetTableName()
+}
+
+// GetFields implements the Primitive interface.
+func (fkc *FkCascade) GetFields(_ context.Context, _ VCursor, _ map[string]*querypb.BindVariable) (*sqltypes.Result, error) {
+	return nil, vterrors.Errorf(vtrpcpb.Code_INTERNAL, "[BUG] GetFields should not be called")
```

Comment:
why not?

Reply:
There is nothing to return on `GetFields` on a DML.
prepare statement ignore calling into `GetFields` for DMLs